### DISCUSSION TITLE
Using `DefinePlugin` to define `import` throws an error

### DISCUSSION THREAD

**Discussion by connor4312** [webpack/webpack#15365] [Q&A]
# Bug report

**What is the current behavior?**

Have a webpack config that includes this in its plugins:

```js
			new DefinePlugin({
				// defined to access the native browser import, which is ordinarily
				// prevented by (1) TS compiling import to AMD require(), and (2) 
				// to its own native module loading. This bypasses both.
				__non_webpack_import__: 'import',
			}),
```

Upon compilation, you'll get an error

```
SyntaxError: Unexpected token (1:7)
    at Object.pp$4.raise (C:\Users\Connor\Documents\Github\vscode-dev\client\node_modules\acorn\dist\acorn.js:3460:15)
    at Object.pp$9.unexpected (C:\Users\Connor\Documents\Github\vscode-dev\client\node_modules\acorn\dist\acorn.js:756:10)
    at Object.pp$5.parseExprImport (C:\Users\Connor\Documents\Github\vscode-dev\client\node_modules\acorn\dist\acorn.js:2854:12)
    at Object.pp$5.parseExprAtom (C:\Users\Connor\Documents\Github\vscode-dev\client\node_modules\acorn\dist\acorn.js:2829:21)
    at Object.pp$5.parseExprSubscripts (C:\Users\Connor\Documents\Github\vscode-dev\client\node_modules\acorn\dist\acorn.js:2627:21)
    at Object.pp$5.parseMaybeUnary (C:\Users\Connor\Documents\Github\vscode-dev\client\node_modules\acorn\dist\acorn.js:2593:19)
    at Object.pp$5.parseExprOps (C:\Users\Connor\Documents\Github\vscode-dev\client\node_modules\acorn\dist\acorn.js:2520:21)
    at Object.pp$5.parseMaybeConditional (C:\Users\Connor\Documents\Github\vscode-dev\client\node_modules\acorn\dist\acorn.js:2503:2
```

**If the current behavior is a bug, please provide the steps to reproduce.**

See above

**What is the expected behavior?**

It should not throw an error.

This is due to the recursion check in the `DefinePlugin`

![image](https://user-images.githubusercontent.com/2230985/153305953-366a2120-955b-4f34-a36b-a68bb5542954.png)

...the JavaScript evaluator tries to parse it as `(import)`, which is not valid.

![image](https://user-images.githubusercontent.com/2230985/153306028-e3da6479-8916-4f32-b425-f8fb81058471.png)

Perhaps if the parsing fails, the code should be emitted as-is?

**Other relevant information:**
webpack version:  5.68.0
Node.js version: v16.13.2
Operating System: Windows 10
Additional tools: N/A

**Comment by vankop** [comment]
`import` is a reserved word, not sure that this must works. you can use/define your own `require`

**Reply by connor4312** [reply]
In this case, I need to use the native browser `import` -- it can't be emulated with my own `require`.

**Reply by alexander-akait** [reply]
Sorry, it is impossible, if you want to inject `import` in code please create loader

**Reply by connor4312** [reply]
That's the workaround I ended up doing, but isn't the original issue still a bug in Webpack? At minimum it should have been emitted as a clear error and not have me debug into Webpack internals to figure out what's going on. Reading the docs it seems like the `DefinePlugin` should work fine for this use case.

**Reply by vankop** [reply]
> That's the workaround I ended up doing, but isn't the original issue still a bug in Webpack?

webpack inject code from DefinePlugin after compilation, not before. So in your case you end up with invalid JS syntax.

maybe `webpackIgnore` comment will be enough for you?

**Reply by connor4312** [reply]
The thing is that the injected code _doesn't_ produce invalid syntax in the compiled code. It's being used as `__non_webpack_import__('./foo.js')` should be replaced to `import('./foo.js')` -- that's valid. The problem is that the way the DefinePlugin is parsing expressions causes it to throw an error.

**Reply by alexander-akait** [reply]
@connor4312 It is parser error, what is error your expected? Error from acorn, so if you want to improve error message, you can send a PR to acorn, we can't fix it here

**Reply by alexander-akait** [reply]
In this case you need to create webpack plugin and register `__non_webpack_import__` as webpack API and then do code generation with `import`, DefinePlugin was not design for this, like we do for `__non_webpack_require__`  https://github.com/webpack/webpack/blob/04039ca99c9a5575ecb041796c6a0858d7be591d/lib/APIPlugin.js#L54

**Reply by vankop** [reply]
>The thing is that the injected code doesn't produce invalid syntax in the compiled code

hm.. in issue description there is syntax error from acorn..

**Reply by connor4312** [reply]
I mean, it shouldn't be a parser error. As a consumer of webpack, I define that `__non_webpack_import__` is equivalent to import, and then expect to be able to use it like I would `import`. None of _my own_ code is invalid or would result in a parse error. It's an implementation detail of Webpack that -- as I've discovered today -- `DefinePlugin` can only be used with valid JS expressions.

I suggest either (a) removing this restriction, or (b) dealing with parse errors and presenting them to the user in a better way. For example in the latter case, Webpack could re-throw an error with a message like `ParseError: expression "import" from the DefinePlugin is not a valid expression`. That would document this behavior and allow consumers to figure out what's going wrong without having to debug webpack internals like I did.

Though, as mentioned, the code emitted from the compiler after the replacement happens would be valid, it's a limitation of the way DefinePlugin is implemented that the plugin couldn't make the replacement.

**Reply by alexander-akait** [reply]
> (b) dealing with parse errors and presenting them to the user in a better way. For example in the latter case, Webpack could re-throw an error with a message like ParseError: expression "import" from the DefinePlugin is not a valid expression.

This error from acorn, feel free to send a PR to them, we can't fix it here as I described above, we can't analyze your code, it will be perf problem - double parsing

**Reply by vankop** [reply]
I see this is an error in evaluating inside `DefinePlugin`. yeah, lets create an issue.

**Reply by vankop** [reply]
https://github.com/webpack/webpack/issues/15371
